java-checks/src/main/java/org/sonar/java/checks/spring/NonSingletonAutowiredInSingletonCheck.java

```diff
@@ -0,0 +1,171 @@
+  }
+
+  private static boolean isNotSingletonAssignmentValue(Arguments arguments) {
+    return arguments
+      .stream()
+      .filter(argument -> argument.is(Tree.Kind.ASSIGNMENT))
+      .map(AssignmentExpressionTree.class::cast)
+      .anyMatch(NonSingletonAutowiredInSingletonCheck::isNotAssignmentToSingletonValue);
+  }
+
+  private static boolean isNotAssignmentToSingletonValue(AssignmentExpressionTree assignmentExpressionTree) {
+    var expression = assignmentExpressionTree.expression();
+
+    if (expression.is(Tree.Kind.STRING_LITERAL)) {
+      return isNotSingletonLiteral(((LiteralTree) expression).value());
+    }
+
+    var variable = (IdentifierTree) assignmentExpressionTree.variable();
+
+    return ("value".equals(variable.name()) || "scopeName".equals(variable.name()))
+      && (expression.is(Tree.Kind.MEMBER_SELECT) && isNotSingletonLiteral(((MemberSelectExpressionTree) expression).identifier().name()));
+  }
+
+  private static boolean isNotSingletonLiteral(String value) {
+    return SINGLETON_LITERALS.stream().noneMatch(singletonLiteral -> singletonLiteral.equalsIgnoreCase(value));
+  }
+
+  private  void analyzeInjectedBeans(Stream<VariableTree> injectedBeans) {
+    injectedBeans
+      .filter(NonSingletonAutowiredInSingletonCheck::hasNotSingletonBeanType)
+      .forEach(variable -> reportIssue(variable.type(), "Singleton beans should not auto-wire non-Singleton beans."));
+  }
+
+  private static boolean hasNotSingletonBeanType(VariableTree classField) {
+    var typeSymbolDeclaration = classField.type().symbolType().symbol().declaration();
+    return typeSymbolDeclaration != null && hasNotSingletonScopeAnnotation(typeSymbolDeclaration.modifiers().annotations());
+  }
+
+  private static Stream<VariableTree> autowiredClassFields(ClassTree classTree) {
+    return classTree.members()
+      .stream()
+      .filter(member -> member.is(Tree.Kind.VARIABLE))
+      .map(VariableTree.class::cast)
+      .filter(classField -> hasAutowiredAnnotation(classField.modifiers().annotations()));
+  }
+
+  private static Stream<VariableTree> parametersFromAutowiredMethods(ClassTree classTree) {
+    return classTree.members()
+      .stream()
+      .filter(member -> member.is(Tree.Kind.METHOD))
+      .map(MethodTree.class::cast)
+      .filter(method -> hasAutowiredAnnotation(method.modifiers().annotations()))
+      .flatMap(methodTree -> methodTree.parameters().stream());
+  }
+
+  private static Stream<VariableTree> parametersFromAutowiredConstructors(ClassTree classTree) {
+    return classTree.members()
+      .stream()
+      .filter(member -> member.is(Tree.Kind.CONSTRUCTOR))
+      .map(MethodTree.class::cast)
+      .filter(NonSingletonAutowiredInSingletonCheck::isAutowiredConstructor)
+      .flatMap(methodTree -> methodTree.parameters().stream());
+  }
+
+  private static Stream<VariableTree> autowiredParametersFromMethodsAndConstructors(ClassTree classTree) {
```

Comment:
What about methods like the following in a controller (random example):
```
  @GetMapping("/hello")
  public Map<String, String> hello(@Autowired PrototypeBean1 fooBean) {
    return Map.of();
  }
```
Shouldn't this be fine? Currently the rule raises an issue here, I think this is an FP. (Happy to discuss!)

Reply:
So, the current implementation will raise an issue. The @Autowired is allowed, even if it should be logically incorrect and thus a potential new rule!

My understanding is that the `@Autowired` is basically used by the ApplicationContext to instantiate objects in those entry points. Your example requires in a way "explicit" invocation from the user, so the `@Autowired` doesn't make any sense. 

Anyway, good point and it is worth more investigation!